tools/btc_proof_generator/src/main.rs

```diff
@@ -0,0 +1,143 @@
+use hex::FromHex;
+use serde::{Deserialize, Serialize};
+
+fn get_merkle_proof(block: &Block, index: usize) -> Result<Vec<Vec<u8>>> {
+    let tx_len = block.txdata.len();
+    assert!(tx_len >= 1);
+    if tx_len == 1 {
+        return Ok(vec![block.txdata[0].txid().to_vec()]);
+    }
+    let mut layers = Vec::new();
+    let mut proof = Vec::new();
+    let layer: Vec<_> = block
+        .txdata
+        .iter()
+        .map(|tx| tx.txid().as_ref().to_vec())
+        .collect();
+    layers.push(layer);
+    let mut current_layer_index = 0;
+    let mut proof_index_in_layer = index;
+    loop {
+        let mut current_layer = layers[current_layer_index].iter();
+        let current_layer_len = current_layer.len();
+        if current_layer_len <= 1 {
+            break;
+        }
+        let new_layer_len = current_layer_len / 2 + current_layer_len % 2;
+        let mut new_layer = Vec::with_capacity(new_layer_len);
+        while let Some(hash1) = current_layer.next() {
+            let hash2 = current_layer.next().unwrap_or(hash1);
+            new_layer.push(hash256_merkle_step(hash1, hash2).as_ref().to_vec());
+        }
+        layers.push(new_layer);
+        let current_layer = layers[current_layer_index].clone();
+        if proof_index_in_layer % 2 == 0 {
+            proof.push(current_layer[proof_index_in_layer + 1].clone());
+        } else {
+            proof.push(current_layer[proof_index_in_layer - 1].clone());
+        }
+        proof_index_in_layer >>= 1;
+        current_layer_index += 1;
+    }
+    Ok(proof)
+}
+
+fn fetch_block(block_hash: &str) -> Result<Block> {
+    let url = format!("https://blockchain.info/block/{}?format=hex", block_hash);
+    let resp = reqwest::blocking::get(&url)?.text()?;
+    let raw_block = Vec::from_hex(resp)?;
+    let block: Block = deserialize(&raw_block)?;
+    Ok(block)
+}
+
+#[derive(Serialize, Deserialize, Default)]
+pub struct MintXTProof {
+    pub version: String,
+    pub vin: String,
+    pub vout: String,
+    pub locktime: String,
+    pub tx_id: String,
+    pub index: u64,
+    pub headers: String,
+    pub intermediate_nodes: String,
+    pub funding_output_index: u8,
+}
+
+fn generate_mint_xt_proof(
+    block_hash: &str,
+    tx_index: usize,
+    funding_output_index: u8,
+) -> Result<(MintXTProof, Block)> {
+    let block = fetch_block(block_hash)?;
+    let tx = block.txdata[tx_index].clone();
+    let proof = get_merkle_proof(&block, tx_index)?;
+    let flat_proof = proof
+        .into_iter()
+        .flat_map(|p| p.into_iter())
+        .collect::<Vec<u8>>();
+
+    Ok((
+        MintXTProof {
+            version: serialize_hex(&tx.version),
+            vin: serialize_hex(&tx.input),
+            vout: serialize_hex(&tx.output),
+            locktime: serialize_hex(&tx.lock_time),
+            tx_id: hex::encode(tx.txid().as_ref()),
+            index: tx_index as u64,
+            headers: serialize_hex(&block.header),
```

Comment:
`headers` field supposed to be current header or with 6 subsequent headers?

it seems current header from code

Reply:
it depends on `TX_PROOF_DIFFICULTY_FACTOR` in config. Right now it is 1, which means only 1 difficulty header can be the proof, so the test contains only one header. If it is 6, this will be 6 raw headers joined together.